src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/FieldSearchHelper.kt

```diff
@@ -52,50 +44,37 @@ internal object FieldSearchHelper {
         data object FoundInNonFinalField: TraverseResult
     }
 
-    private fun findObjectField(testObject: Any?, value: Any, visitedObjects: MutableSet<Any>): TraverseResult {
+    private fun findObjectField(testObject: Any?, value: Any): TraverseResult {
         if (testObject == null) return NotFound
-        var fieldName: OwnerWithName? = null
-        // We take all the fields from the hierarchy.
-        // If two or more fields match (===) the AtomicReference object, we fall back to the default behavior,
-        // so there is no problem that we can receive some fields of the same name and the same type.
-        for (field in testObject::class.java.allDeclaredFieldWithSuperclasses) {
-            if (field.type.isPrimitive) continue
-
-            // We wrap an unsafe read into `runCatching` to hande `UnsupportedOperationException`,
-            // which can be thrown, for instance, when attempting to read a field of
-            // a hidden class (starting from Java 15).
-            val fieldValue = runCatching {
-                readFieldViaUnsafe(testObject, field, Unsafe::getObject)
-            }.getOrNull() ?: continue
 
-            if (fieldValue in visitedObjects) continue
-            visitedObjects += testObject
+        var traverseResult: TraverseResult = NotFound
+        var fieldName: OwnerWithName? = null
+        val isTraverseCompleted = {
+            traverseResult is MultipleFieldsMatching ||
+            traverseResult is FoundInNonFinalField
+        }
 
-            if (fieldValue === value) {
-                if (fieldName != null) return MultipleFieldsMatching
-                if (!Modifier.isFinal(field.modifiers)) return FoundInNonFinalField
+        traverseObjectHierarchy(testObject, treatArrayElementsAsFields = false) { ownerObject, field, fieldValue ->
+            if (field.type.isPrimitive || fieldValue == null) return@traverseObjectHierarchy null
 
-                fieldName = if (Modifier.isStatic(field.modifiers)) {
-                    StaticOwnerWithName(field.name, testObject::class.java)
-                } else {
-                    InstanceOwnerWithName(field.name, testObject)
-                }
-                continue
-            }
-            when (val result = findObjectField(fieldValue, value, visitedObjects)) {
-                is FieldName -> {
-                    if (fieldName != null) {
-                        return MultipleFieldsMatching
+            if (value === fieldValue && !isTraverseCompleted()) {
+                if (fieldName != null) traverseResult = MultipleFieldsMatching
```

Comment:
Please put `traverseResult = MultipleFieldsMatching` into parentheses `{ traverseResult = MultipleFieldsMatching } ` for clarity.